File: prototype/solver.py
"""Rescale dimension solver — prototype.

Model: the target size has exactly 2 degrees of freedom (width, height).
Everything the user can type -- aspect, megapixels, width, height, long edge,
short edge, scale -- is a *constraint* on those two. The user pins constraints
by editing them; the two most-recently-edited independent ones win, and any
remaining freedom is inherited from the source image.

`multiple` is not a degree of freedom. It is a quantisation applied after the
continuous solve, choosing the nearby lattice point with the least aspect-ratio
deviation.
"""
import math
from dataclasses import dataclass, field

# aspect ratio deviation dominates; pixel-count deviation only breaks ties.
W_ASPECT, W_PIXELS = 1.0, 0.05


@dataclass
class Solution:
    width: int
    height: int
    ideal_w: float
    ideal_h: float
    aspect_error_pct: float
    pixel_error_pct: float
    used: list = field(default_factory=list)


def solve_continuous(pins, src_w, src_h):
    """pins: ordered most-recent-first list of (key, value). Returns (w, h, used)."""
    src_ar, src_px = src_w / src_h, src_w * src_h
    c, used = {}, []
    for key, val in pins:
        if key in c:
            continue
        trial = dict(c)
        trial[key] = val
        if _rank(trial) > _rank(c):          # keep only constraints that add information
            c, _ = trial, used.append(key)
        if _rank(c) == 2:
            break

    # inherit remaining freedom from the source
    if _rank(c) < 2:
        if "aspect" not in c and _rank({**c, "aspect": src_ar}) > _rank(c):
            c["aspect"] = src_ar
    if _rank(c) < 2:
        c["megapixels"] = src_px / 1e6

    return (*_solve(c, src_ar, src_px), used)


def _rank(c):
    """How many degrees of freedom this constraint set removes (0, 1 or 2)."""
    n = len(c)
    return 2 if n >= 2 else n


def _solve(c, src_ar, src_px):
    ar = c.get("aspect")
    if "width" in c and "height" in c:
        return float(c["width"]), float(c["height"])
    if "scale" in c:                                  # linear scale of the source
        c = {**c, "megapixels": src_px * c["scale"] ** 2 / 1e6}
        c.pop("scale")
    if ar is not None:
        if "megapixels" in c:
            px = c["megapixels"] * 1e6
            return math.sqrt(px * ar), math.sqrt(px / ar)
        if "width" in c:
            return float(c["width"]), c["width"] / ar
        if "height" in c:
            return c["height"] * ar, float(c["height"])
        if "long" in c:
            L = c["long"]
            return (float(L), L / ar) if ar >= 1 else (L * ar, float(L))
        if "short" in c:
            S = c["short"]
            return (S * ar, float(S)) if ar >= 1 else (float(S), S / ar)
    if "megapixels" in c:
        px = c["megapixels"] * 1e6
        if "width" in c:
            return float(c["width"]), px / c["width"]
        if "height" in c:
            return px / c["height"], float(c["height"])
    raise ValueError(f"unsolvable constraint set: {c}")


def snap(w, h, multiple, target_ar, span=2):
    """Pick the lattice point near (w, h) with the least aspect deviation."""
    if multiple <= 1:
        return max(1, round(w)), max(1, round(h))
    target_px = w * h
    base_w, base_h = math.floor(w / multiple), math.floor(h / multiple)
    best, best_score = None, float("inf")
    for i in range(-span, span + 2):
        for j in range(-span, span + 2):
            cw, ch = (base_w + i) * multiple, (base_h + j) * multiple
            if cw < multiple or ch < multiple:
                continue
            score = (W_ASPECT * math.log((cw / ch) / target_ar) ** 2
                     + W_PIXELS * math.log((cw * ch) / target_px) ** 2)
            if score < best_score:
                best, best_score = (cw, ch), score
    return best


def solve(pins, src_w, src_h, multiple=1):
    w, h, used = solve_continuous(pins, src_w, src_h)
    target_ar = w / h
    sw, sh = snap(w, h, multiple, target_ar)
    return Solution(sw, sh, w, h,
                    abs((sw / sh) / target_ar - 1) * 100,
                    abs((sw * sh) / (w * h) - 1) * 100, used)

# ---------------------------------------------------------------------------
MULTIPLES = (1, 8, 16)


def anchored_axes(pins, w, h):
    """Which axes did the user pin *numerically*? Those must survive snapping."""
    a = set()
    for key, _ in pins:
        if key == "width":    a.add("w")
        elif key == "height": a.add("h")
        elif key == "long":   a.add("w" if w >= h else "h")
        elif key == "short":  a.add("h" if w >= h else "w")
    return a


def snap_respecting_pins(w, h, multiple, target_ar, anchors, span=3):
    """Anchored axes snap to the nearest lattice value and are then held fixed;
    only free axes are searched."""
    if multiple <= 1:
        return max(1, round(w)), max(1, round(h))
    target_px = w * h

    def lattice(v):
        return max(multiple, round(v / multiple) * multiple)

    def window(v):
        b = math.floor(v / multiple)
        return [(b + i) * multiple for i in range(-span, span + 2)
                if (b + i) * multiple >= multiple]

    ws = [lattice(w)] if "w" in anchors else window(w)
    hs = [lattice(h)] if "h" in anchors else window(h)

    best, best_score = None, float("inf")
    for cw in ws:
        for ch in hs:
            score = (W_ASPECT * math.log((cw / ch) / target_ar) ** 2
                     + W_PIXELS * math.log((cw * ch) / target_px) ** 2)
            if score < best_score:
                best, best_score = (cw, ch), score
    return best


def hold_anchors(w, h, multiple, anchors):
    """Snap the anchored axis to the lattice first, then re-derive the free axis
    from the value the user will actually get. Without this, a typed 1500 at
    16:9 would pick a height that best matches 1500 -- a width the user is not
    getting -- rather than the 1504 that is being written."""
    if multiple <= 1:
        return w, h
    ar = w / h
    if "w" in anchors and "h" not in anchors:
        w2 = max(multiple, round(w / multiple) * multiple)
        return w2, w2 / ar
    if "h" in anchors and "w" not in anchors:
        h2 = max(multiple, round(h / multiple) * multiple)
        return h2 * ar, h2
    return w, h


def solve2(pins, src_w, src_h, multiple=1):
    w, h, used = solve_continuous(pins, src_w, src_h)
    anchors = anchored_axes(pins, w, h)
    w, h = hold_anchors(w, h, multiple, anchors)
    target_ar = w / h
    sw, sh = snap_respecting_pins(w, h, multiple, target_ar, anchors)
    return Solution(sw, sh, w, h,
                    abs((sw / sh) / target_ar - 1) * 100,
                    abs((sw * sh) / (w * h) - 1) * 100, used)

File: RescaleKit/Sources/RescaleKit/Solver.swift
import Foundation

/// Everything the user can set about the target geometry.
public struct ResizeRequest: Hashable, Sendable, Codable {
    public var aspect: AspectRatio
    public var size: SizeParameter
    public var multiple: Multiple

    public init(aspect: AspectRatio = .original, size: SizeParameter, multiple: Multiple = .one) {
        self.aspect = aspect
        self.size = size
        self.multiple = multiple
    }

    public func solve(for source: PixelSize) -> Solution {
        Solver.solve(self, source: source)
    }
}

/// The dimension solver, ported from `prototype/solver.py` (`solve2`).
///
/// Two stages:
/// 1. **Continuous solve.** Aspect + one size parameter give an exact real-valued
///    (W*, H*). There is no competition between constraints, so no pins to
///    arbitrate — the aspect is always known and the size is one number.
/// 2. **Pin-respecting snap.** The axis the user typed snaps to its own nearest
///    lattice value and is then held fixed; only the derived axis is searched
///    for the lattice point with least aspect deviation, ties broken by pixel
///    count (PRD §6).
public enum Solver {
    /// Aspect-ratio deviation dominates; pixel-count deviation only breaks ties.
    static let aspectWeight = 1.0
    static let pixelWeight = 0.05
    /// How many lattice steps either side of the ideal a free axis is searched.
    static let searchSpan = 3

    public static func solve(_ request: ResizeRequest, source: PixelSize) -> Solution {
        precondition(source.width > 0 && source.height > 0, "source must have positive dimensions")

        let pinned = request.size.pinnedAxis
        let typed = solveContinuous(request, source: source)
        let ideal = holdPinnedAxis(typed, multiple: request.multiple.rawValue, pinned: pinned)
        let size = snap(ideal, multiple: request.multiple.rawValue, pinned: pinned)

        return Solution(
            size: size,
            ideal: ideal,
            pinnedAxis: pinned,
            adjustments: adjustments(request: request, ideal: ideal, size: size)
        )
    }

    // MARK: - Stage 1: continuous solve

    static func solveContinuous(_ request: ResizeRequest, source: PixelSize) -> ContinuousSize {
        let ar = request.aspect.ratio(for: source)
        switch request.size {
        case let .width(w):
            let w = Double(w)
            return ContinuousSize(width: w, height: w / ar)
        case let .height(h):
            let h = Double(h)
            return ContinuousSize(width: h * ar, height: h)
        case let .megapixels(mp):
            return fromPixelCount(mp * Megapixel.pixels, aspect: ar)
        case let .scale(s):
            // A linear scale of the source is a pixel budget of source × s².
            // Routed through megapixels exactly as the prototype does, so the
            // two stay numerically identical.
            let mp = Double(source.pixelCount) * s * s / Megapixel.pixels
            return fromPixelCount(mp * Megapixel.pixels, aspect: ar)
        }
    }

    private static func fromPixelCount(_ px: Double, aspect ar: Double) -> ContinuousSize {
        ContinuousSize(width: (px * ar).squareRoot(), height: (px / ar).squareRoot())
    }

    // MARK: - Stage 2: pin-respecting snap

    /// Snap the pinned axis to the lattice first, then re-derive the free axis
    /// from the value the user will actually get. Typing 1500 at 16:9 with
    /// multiple 8 yields a width of 1504, so the height should be the best
    /// match for 1504, not for 1500. This also makes re-solving with the
    /// solved value a true fixed point.
    static func holdPinnedAxis(_ typed: ContinuousSize, multiple: Int, pinned: Axis?) -> ContinuousSize {
        guard multiple > 1 else { return typed }
        let ar = typed.aspectRatio
        switch pinned {
        case .width:
            let w = Double(nearestLattice(typed.width, multiple))
            return ContinuousSize(width: w, height: w / ar)
        case .height:
            let h = Double(nearestLattice(typed.height, multiple))
            return ContinuousSize(width: h * ar, height: h)
        case nil:
            return typed
        }
    }

    static func snap(_ ideal: ContinuousSize, multiple: Int, pinned: Axis?) -> PixelSize {
        if multiple <= 1 {
            return PixelSize(width: max(1, roundHalfEven(ideal.width)),
                             height: max(1, roundHalfEven(ideal.height)))
        }

        let targetAR = ideal.aspectRatio
        let targetPx = ideal.pixelCount

        let widths = pinned == .width ? [nearestLattice(ideal.width, multiple)] : window(ideal.width, multiple)
        let heights = pinned == .height ? [nearestLattice(ideal.height, multiple)] : window(ideal.height, multiple)

        var best = PixelSize(width: widths[0], height: heights[0])
        var bestScore = Double.infinity
        // Iteration order matters for ties: smallest candidates win, as in the prototype.
        for w in widths {
            for h in heights {
                let cw = Double(w), ch = Double(h)
                let arTerm = log((cw / ch) / targetAR)
                let pxTerm = log((cw * ch) / targetPx)
                let score = aspectWeight * arTerm * arTerm + pixelWeight * pxTerm * pxTerm
                if score < bestScore {
                    best = PixelSize(width: w, height: h)
                    bestScore = score
                }
            }
        }
        return best
    }

    /// The nearest lattice value, never below one step. Half-way cases round to
    /// even (1032 at 16 → 64.5 steps → 64 → 1024), matching the prototype and PRD §6.
    static func nearestLattice(_ v: Double, _ multiple: Int) -> Int {
        max(multiple, roundHalfEven(v / Double(multiple)) * multiple)
    }

    /// Lattice values from `searchSpan` steps below the ideal to `searchSpan + 1` above.
    static func window(_ v: Double, _ multiple: Int) -> [Int] {
        let base = Int((v / Double(multiple)).rounded(.down))
        return (-searchSpan...(searchSpan + 1))
            .map { (base + $0) * multiple }
            .filter { $0 >= multiple }
    }

    static func roundHalfEven(_ v: Double) -> Int {
        Int(v.rounded(.toNearestOrEven))
    }

    // MARK: - Reporting

    static func adjustments(request: ResizeRequest, ideal: ContinuousSize, size: PixelSize) -> [Solution.Adjustment] {
        var out: [Solution.Adjustment] = []
        switch request.size {
        case let .width(w) where w != size.width:
            out.append(.pinnedValueSnapped(axis: .width, requested: w, actual: size.width))
        case let .height(h) where h != size.height:
            out.append(.pinnedValueSnapped(axis: .height, requested: h, actual: size.height))
        default:
            break
        }
        if request.multiple.isIdentity { return out }

        // The derived axis was "forced" when the multiple, not rounding, moved it.
        let idealW = roundHalfEven(ideal.width)
        let idealH = roundHalfEven(ideal.height)
        if request.size.pinnedAxis != .width, idealW != size.width {
            out.append(.derivedAxisForced(axis: .width, ideal: idealW, actual: size.width))
        }
        if request.size.pinnedAxis != .height, idealH != size.height {
            out.append(.derivedAxisForced(axis: .height, ideal: idealH, actual: size.height))
        }
        return out
    }
}
